test_communication/CMakeLists.txt

```diff
@@ -280,6 +306,158 @@ if(BUILD_TESTING)
     endforeach()
   endmacro()
 
+  macro(security_tests)
+    set(suffix "__${rmw_implementation}")
+    set(PUBLISHER_RMW ${rmw_implementation})
+    set(SUBSCRIBER_RMW ${rmw_implementation})
+    # Not testing across client libraries for now
+    set(TEST_PUBLISHER_RCL "rclcpp")
+    set(TEST_SUBSCRIBER_RCL "rclcpp")
+    set(TEST_PUBLISHER_EXECUTABLE "$<TARGET_FILE:test_secure_publisher_cpp>")
+    set(TEST_SUBSCRIBER_EXECUTABLE "$<TARGET_FILE:test_secure_subscriber_cpp>")
+
+    # Test suite for communication without security
+    set(non_secure_comm_PUBLISHER_ROS_ENABLE_SECURITY_LIST "false;false;false;true;false")
+    set(non_secure_comm_SUBSCRIBER_ROS_ENABLE_SECURITY_LIST "false;false;false;false;true")
+    set(non_secure_comm_PUBLISHER_ROS_SECURITY_STRATEGY_LIST "Enforce;garbage;garbage;Permissive;Garbage")
+    set(non_secure_comm_SUBSCRIBER_ROS_SECURITY_STRATEGY_LIST "Enforce;Permissive;Garbage;Garbage;Permissive")
+    set(non_secure_comm_PUBLISHER_ROS_SECURE_ROOT_LIST "garbage;WHATEVER;${VALID_ROS_SECURE_ROOT};garbage;garbage")
+    set(SUBSCRIBER_ROS_SECURE_ROOT_LIST "${VALID_ROS_SECURE_ROOT};WHATEVER;garbage;garbage;garbage")
+
+    # Test suite for secured communication
+    set(secure_comm_PUBLISHER_ROS_ENABLE_SECURITY_LIST "true;true;true;true")
+    set(secure_comm_SUBSCRIBER_ROS_ENABLE_SECURITY_LIST "true;true;true;true")
+    set(secure_comm_PUBLISHER_ROS_SECURITY_STRATEGY_LIST "Enforce;Enforce;Permissive;Permissive")
+    set(secure_comm_SUBSCRIBER_ROS_SECURITY_STRATEGY_LIST "Enforce;Permissive;Enforce;Permissive")
+
+    # Test suite for one node with security and the second without
+    set(not_connecting_PUBLISHER_ROS_ENABLE_SECURITY_LIST "false;true")
+    set(not_connecting_SUBSCRIBER_ROS_ENABLE_SECURITY_LIST "true;false")
+    set(not_connecting_PUBLISHER_ROS_SECURITY_STRATEGY_LIST "Permissive;Enforce")
+    set(not_connecting_SUBSCRIBER_ROS_SECURITY_STRATEGY_LIST "Enforce;Permissive")
+
+    set(n_non_secure_tests "")
+    set(n_secure_communication_tests "")
+    set(n_not_connecting_tests "")
+    list(LENGTH non_secure_comm_PUBLISHER_ROS_ENABLE_SECURITY_LIST n_non_secure_tests)
+    list(LENGTH secure_comm_PUBLISHER_ROS_ENABLE_SECURITY_LIST n_secure_communication_tests)
+    list(LENGTH not_connecting_PUBLISHER_ROS_ENABLE_SECURITY_LIST n_not_connecting_tests)
+
+    foreach(message_file ${message_files})
+      get_filename_component(TEST_MESSAGE_TYPE "${message_file}" NAME_WE)
+      # TODO(mikaelarguedas) test only few message types to avoid taking to much test time
+      if(TEST_MESSAGE_TYPE STREQUAL "Empty" OR TEST_MESSAGE_TYPE STREQUAL "DynamicArrayNested")
+        set(index 0)
+        # configure all non secure communication tests
+        set(SUBSCRIBER_SHOULD_TIMEOUT "false")
+        while(index LESS ${n_non_secure_tests})
```

Comment:
I tried to use the same template for all tests, this result in this nested while loops inside the foreach. Any advice on how we can make this more redable is appreciated